### DISCUSSION TITLE
parse from file with limited memory usage

### DISCUSSION THREAD

**Discussion by ricksapphire** [capnproto/capnproto#1936] [Q&A]
```
struct Metric {
    values @1 : List(List(UInt32)); // at most 268435456 for inner part
}
```

I have a pretty large(above 3GB) file with the schema above and I use mmap with FlatArrayMessageReader to read this file periodically in a c++ server. I notice that the server will stall about 200ms while reading the file due to large memory allocation and release. I try StreamFdMessageReader but it seems use more memory.

In my case, I want to parse the list sequentially. Is there any method to parse the file as stream and parse with limited memory usage?

**Comment by kentonv** [comment]
Typically, mmap() and read()ing a file use the same amount of memory. Either way, the kernel first reads the file contents into page cache. In the mmap() case, the pages in page cache are directly mapped into userspace (this takes ~no additional memory) whereas in the read() case, the contents are copied into userspace (this actually takes slightly _more_ memory because you need to allocate a userspace buffer). In either case, when the kernel needs memory for something else, it will drop the pages from page cache. If they are currently mapped into userspace, it unmaps them.

Are you sure that read()ing your file sequentially wouldn't _also_ take 200ms?

The Cap'n Proto library does not offer any streaming implementation for reading an individual message. But, I would expect mmap to be the best solution here anyway.

**Reply by ricksapphire** [reply]
Actually StreamFdMessageReader will use more memory, but with tcmalloc in the server, it will cache the allocated memory which will result in higher memory occupation, but it will not stall the server for 200ms. 

But in my case, higher memory occupation is also not designable. I'm not very deep into the message format, it is possible to locate the single message in the list, so I can just mmap or read several message once at a time?

**Reply by kentonv** [reply]
`StreamFdMessageReader` reads the entire message into memory all at once, keeping it there until the reader is destroyed. Of course that uses more memory. I was trying to compare between mmap() and a hypothetical solution that uses streaming read()s without keeping everything in memory at once.

With that said, `StreamFdMessageReader` still forces the pages to be brought into page cache in the kernel, as mmap() does. So this approach does all the same memory allocation needed by the mmap() approach, _plus_ the userspace allocation (the latter of which may be cached by tcmalloc).

Have you actually observed a 200ms stall with mmap() but no stall with `StreamFdMessageReader` in real tests? This would surprise me, but if that is the case, then I think something else is going on which needs to be understood.

But in any case, no, the library does not provide a way to read just one element of the list except by mmap()ing the whole message.
